components/txn_types/src/write.rs

```diff
@@ -259,6 +259,25 @@ impl Write {
             versions_to_last_change: self.versions_to_last_change,
         }
     }
+
+    /// Returns the new `last_change_ts` and `versions_to_last_change` according
+    /// to this write record.
+    pub fn next_last_change_info(&self, commit_ts: TimeStamp) -> (TimeStamp, u64) {
+        match self.write_type {
+            WriteType::Put | WriteType::Delete => (commit_ts, 1),
+            WriteType::Lock | WriteType::Rollback => {
+                // If `last_change_ts` is zero, do not set `last_change_ts` to indicate we don't
+                // know where is the last change.
+                // This should not happen if data is written in new version TiKV. If we hope to
+                // support data from old TiKV, consider iterating to the last change to find it.
+                if !self.last_change_ts.is_zero() {
+                    (self.last_change_ts, self.versions_to_last_change + 1)
+                } else {
+                    (TimeStamp::zero(), 0)
```

Comment:
Is it possible that we just return `(commit_ts, 1)` in this case, and if we seek to the `last_change_ts` and find it's Lock or Rollback, do `next`s to find the actual version? I think in this way maybe it's possible that for those keys that are frequently locked but rarely updated, they can benefit from the optimization after upgrading from an old version.

By the way, GC may change the real number of versions to last change. Does that matter?

Reply:
> Is it possible that we just return `(commit_ts, 1)` in this case, and if we seek to the `last_change_ts` and find it's Lock or Rollback when reading, do `next`s to find the actual data? I think in this way maybe it's possible that for those keys that are frequently locked but rarely updated, they can benefit from the optimization after upgrading from an old version.

Emm, actually I don't intend to solve the problem for upgraded data at the first step.

If we indeed want to solve the problem for upgraded version, my idea is to iterate until the last change in `acquire_pessimistic_lock` or `prewrite` to get the correct value (which is what the comment in L272 means). Then, only the first time a key is written will be slow. I think this way is a bit better.

Reply:
> By the way, GC may change the real number of versions to last change (by deleting old `Lock`s and `Rollback`s). Does that matter?

Right... This also brings inaccuracy. I don't have a good idea now but I don't think it's a very serious problem. The performance just doesn't come back to optimism, and it is likely for such a key to pile up enough versions again.

Anyway, it's always pleasant to solve this problem if possible. Do you have any good idea...?

Reply:
> > By the way, GC may change the real number of versions to last change (by deleting old `Lock`s and `Rollback`s). Does that matter?
> 
> Right... This also brings inaccuracy. I don't have a good idea now but I don't think it's a very serious problem. The performance just doesn't come back to optimism, and it is likely for such a key to pile up enough versions again.
> 
> Anyway, it's always pleasant to solve this problem if possible. Do you have any good idea...?

I have no other idea either. I think it's ok to ignore the inaccuracy too. (I'm just confirming that you already noticed this case